OrbitCore/LinuxPerfUtils.h

```diff
@@ -0,0 +1,124 @@
+//-----------------------------------
+// Copyright Pierric Gimmig 2013-2019
+//-----------------------------------
+#pragma once
+
+#include <asm/unistd.h>
+#include <asm/perf_regs.h>
+#include <linux/perf_event.h>
+#include <sys/errno.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+//-----------------------------------------------------------------------------
+namespace LinuxPerfUtils
+{
+    // TODO: these constants should be made available in some settings
+    static constexpr uint32_t STACK_SIZE = (1u << 13u) - 8; // TODO: the stack is not yet recorded
+    static constexpr size_t RING_BUFFER_PAGE_COUNT = 2048; // 64: 256 KB; 2048: 8 MB
+    // sample stack and instruction pointer (used later for stack unwinding)
+    static constexpr uint64_t SAMPLE_REGS_USER = (0x1 << PERF_REG_X86_SP) | (0x1 << PERF_REG_X86_IP);
+
+    // This must be in sync with the perf_event_record and the perf_sample_id stucts.
+    static constexpr uint64_t SAMPLE_TYPE_FLAGS = 
+        PERF_SAMPLE_TID | 
+        PERF_SAMPLE_TIME | 
+        PERF_SAMPLE_CPU | 
+        0u;
+
+    typedef int NullType[0];
+
+    // This struct must be in sync with the SAMPLE_TYPE_FLAGS.
+    struct perf_sample_id {
+        uint32_t pid, tid;  /* if PERF_SAMPLE_TID set */
+        uint64_t time;      /* if PERF_SAMPLE_TIME set */
+        uint32_t cpu, res;  /* if PERF_SAMPLE_CPU set */
+    };
+
+    struct perf_event_fork_exit {
+        struct perf_event_header header;
+        uint32_t    pid, ppid;
+        uint32_t    tid, ptid;
+        uint64_t    time;
+        struct perf_sample_id sample_id;
+    };
+
+    struct perf_event_lost {
+        struct perf_event_header header;
+        uint64_t    id;
+        uint64_t    lost;
+        struct perf_sample_id sample_id;
+    };
+
+    // This struct must be in sync with the SAMPLE_REGS_USER.
+    struct perf_sample_regs_user {
+        /* if PERF_SAMPLE_REGS_USER */
+        uint64_t    abi;
+        // uint64_t    regs[__builtin_popcount(SAMPLE_REGS_USER)]; 
+        uint64_t sp;
+        uint64_t ip;
+    };
+
+    struct perf_sample_stack_user {
+        uint64_t    size;        /* if PERF_SAMPLE_STACK_USER */
+        char   data[STACK_SIZE]; /* if PERF_SAMPLE_STACK_USER */
+        uint64_t    dyn_size;    /* if PERF_SAMPLE_STACK_USER && size != 0 */
+    };
+
+    // This struct must be in sync with the SAMPLE_TYPE_FLAGS.
+    template<typename raw_data_t, typename registers_t, typename stack_t>
+    struct __attribute__((__packed__)) perf_event_record {
```

Comment:
Can we somehow implement the "accept" method here such that we don't need to copy twice?

Reply:
I just pushed a solution, where I use the LinuxPerfEvent (sub)classes directly as a target for the memcpy from the ringbuffer.